File: k_means/matlab/kmeans.m
classdef Point
    properties
        x
        y
    end
    
    methods
        function obj = Point(x, y)
            obj.x = x;
            obj.y = y;
        end
    end
end

classdef Cluster
    properties
        centroid
        numPoints
    end
    
    methods
        function obj = Cluster(centroid)
            obj.centroid = centroid;
            obj.numPoints = 0;
        end
    end
end

function dist = distance(p1, p2)
    dist = sqrt((p2.x - p1.x)^2 + (p2.y - p1.y)^2);
end

function centroids = initializeCentroids(dataPoints, k)
    centroids = Point.empty;
    for i = 1:k
        randomIndex = randi(length(dataPoints));
        centroids(i) = dataPoints(randomIndex);
    end
end

function assignPointsToCentroids(dataPoints, clusters)
    for i = 1:length(dataPoints)
        minDistance = distance(dataPoints(i), clusters(1).centroid);
        clusterIndex = 1;
        
        for j = 2:length(clusters)
            currDistance = distance(dataPoints(i), clusters(j).centroid);
            if currDistance < minDistance
                minDistance = currDistance;
                clusterIndex = j;
            end
        end
        
        clusters(clusterIndex).numPoints = clusters(clusterIndex).numPoints + 1;
    end
end

function updateCentroids(dataPoints, clusters)
    for i = 1:length(clusters)
        sumX = 0.0;
        sumY = 0.0;
        numPoints = clusters(i).numPoints;
        
        for j = 1:length(dataPoints)
            sumX = sumX + dataPoints(j).x;
            sumY = sumY + dataPoints(j).y;
        end
        
        clusters(i).centroid.x = sumX / numPoints;
        clusters(i).centroid.y = sumY / numPoints;
    end
end

function clusters = kMeans(dataPoints, k)
    clusters = Cluster.empty;
    for i = 1:k
        clusters(i) = Cluster(Point(0, 0));
    end
    
    iteration = 0;
    maxIterations = 100;
    while iteration < maxIterations
        assignPointsToCentroids(dataPoints, clusters);
        updateCentroids(dataPoints, clusters);
        iteration = iteration + 1;
    end
end

dataPoints = [Point(2.0, 3.0), Point(2.5, 4.5), Point(1.5, 2.5), Point(6.0, 5.0), Point(7.0, 7.0), Point(5.0, 5.5), Point(9.0, 2.0), Point(10.0, 3.5), Point(9.5, 2.5)];
k = 3;

clusters = kMeans(dataPoints, k);

for i = 1:length(clusters)
    fprintf("Cluster %d: Centroid (%.2f, %.2f), Points: %d\n", i, clusters(i).centroid.x, clusters(i).centroid.y, clusters(i).numPoints);
end

File: k_means/rust/kmeans.rs
use rand::seq::SliceRandom;
use rand::thread_rng;
use std::f64;

#[derive(Clone, Copy)]
struct Point {
    x: f64,
    y: f64,
}

impl Point {
    fn new(x: f64, y: f64) -> Self {
        Point { x, y }
    }
}

struct Cluster {
    centroid: Point,
    num_points: u32,
}

impl Cluster {
    fn new(centroid: Point) -> Self {
        Cluster {
            centroid,
            num_points: 0,
        }
    }
}

fn distance(p1: Point, p2: Point) -> f64 {
    ((p2.x - p1.x).powi(2) + (p2.y - p1.y).powi(2)).sqrt()
}

fn initialize_centroids(data_points: &[Point], k: usize) -> Vec<Point> {
    let mut rng = thread_rng();
    let mut centroids: Vec<Point> = data_points
        .choose_multiple(&mut rng, k)
        .cloned()
        .collect();
    centroids
}

fn assign_points_to_centroids(data_points: &[Point], clusters: &mut Vec<Cluster>) {
    for point in data_points {
        let mut min_distance = f64::MAX;
        let mut cluster_index = 0;
        for (j, cluster) in clusters.iter().enumerate() {
            let curr_distance = distance(*point, cluster.centroid);
            if curr_distance < min_distance {
                min_distance = curr_distance;
                cluster_index = j;
            }
        }
        clusters[cluster_index].num_points += 1;
    }
}

fn update_centroids(data_points: &[Point], clusters: &mut Vec<Cluster>) {
    for cluster in clusters.iter_mut() {
        let mut sum_x = 0.0;
        let mut sum_y = 0.0;
        let num_points = cluster.num_points as f64;

        for point in data_points {
            sum_x += point.x;
            sum_y += point.y;
        }

        cluster.centroid.x = sum_x / num_points;
        cluster.centroid.y = sum_y / num_points;
    }
}

fn k_means(data_points: &[Point], k: usize) -> Vec<Cluster> {
    let mut clusters: Vec<Cluster> = initialize_centroids(data_points, k)
        .into_iter()
        .map(Cluster::new)
        .collect();

    let max_iterations = 100;
    let mut iteration = 0;

    while iteration < max_iterations {
        assign_points_to_centroids(data_points, &mut clusters);
        update_centroids(data_points, &mut clusters);
        iteration += 1;
    }

    clusters
}

fn main() {
    let data_points = [
        Point::new(2.0, 3.0),
        Point::new(2.5, 4.5),
        Point::new(1.5, 2.5),
        Point::new(6.0, 5.0),
        Point::new(7.0, 7.0),
        Point::new(5.0, 5.5),
        Point::new(9.0, 2.0),
        Point::new(10.0, 3.5),
        Point::new(9.5, 2.5),
    ];

    let k = 3;

    let clusters = k_means(&data_points, k);

    for (i, cluster) in clusters.iter().enumerate() {
        println!(
            "Cluster {}: Centroid ({:.2}, {:.2}), Points: {}",
            i + 1,
            cluster.centroid.x,
            cluster.centroid.y,
            cluster.num_points
        );
    }
}